File: LeetCode Solutions/Learn/Binary Tree/Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if(not root):
            return False
        def helper(node, total):
            if(node):
                if(not node.left and not node.right):
                    return total + node.val == sum
                return helper(node.left, total + node.val) or helper(node.right, total + node.val)
        return helper(root, 0)

File: LeetCode Solutions/October 2022 LeetCoding Challenge/Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int x = 4;
    public void dfs(TreeNode n, int sum){
        if (n == null){
            return;
        }
        if (n.left == null & n.right == null){
            if (sum-n.val == 0){
                x = 0;
            }
        } else{
            dfs(n.left, sum-n.val);
            dfs(n.right, sum-n.val);
        }
    }
    public boolean hasPathSum(TreeNode root, int totalSum) {
        dfs(root, totalSum);
        return x == 0;
    }
}